src/core/algorithms/fd/pfdtane/pfd_verifier/pfd_stats_calculator.h

```diff
@@ -0,0 +1,98 @@
+#pragma once
+
+#include "config/error/type.h"
+#include "config/indices/type.h"
+#include "config/error_measure/type.h"
+#include "config/indices/type.h"
+#include "model/table/column_layout_relation_data.h"
+#include "model/table/position_list_index.h"
+
+namespace algos {
+
+class PFDStatsCalculator {
+private:
+    std::shared_ptr<ColumnLayoutRelationData> relation_;
+    config::ErrorType max_fd_error_;
+    config::ErrorMeasureType error_measure_;
+
+    std::vector<model::PLI::Cluster> clusters_violating_pfd_;
+    size_t num_rows_violating_pfd_ = 0;
+    config::ErrorType error_ = 0.0;
+
+public:
+    explicit PFDStatsCalculator(std::shared_ptr<ColumnLayoutRelationData> relation,
+                                config::ErrorMeasureType measure, config::ErrorType max_fd_error)
+        : relation_(std::move(relation)), max_fd_error_(max_fd_error), error_measure_(measure) {}
+
+    void ResetState() {
+        clusters_violating_pfd_.clear();
+        num_rows_violating_pfd_ = 0;
+        error_ = 0;
+    }
+
+    bool PFDHolds() const {
```

Comment:
Not sure if we need a separate function to check whether pfd holds or not (and store `max_fd_error_`) since `afd_verifier` just returns the error to the user and pfd is pretty much an afd. But this is a topic to discuss with the maintainers, let me know if you've done it already.

Reply:
For the future reference: we just talked and decided to fix immediate problems, found by @egshnov. The architecture rework will come later, when both mining and validation PRs for other metrics will be merged.